src/redteam/commands/discover.ts

```diff
@@ -144,19 +104,50 @@ export async function doTargetPurposeDiscovery(
       method: 'POST',
     });
 
-    const { purpose } = (await res.json()) as { purpose?: string };
+    const { done, purpose, question } = (await res.json()) as {
+      question?: string;
+      done: boolean;
+      purpose?: string;
+    };
+
+    // `done` will be true if max turns are reached (server is aware of the turn count) OR the server
+    // is satisfied with the current purpose.
+    if (
+      done ||
+      // `done` should never not be `true` here, but it's good to have a fallback:
+      i === maxTurns
+    ) {
+      // Finalize the progress bar:
+      pbar.increment();
+      pbar.stop();
 
-    if (purpose) {
-      return purpose;
-    }
+      // Handle the edge case where the server did not call `done` when max turns are reached:
+      if (purpose) {
+        logger.info(`\nPurpose:\n\n${chalk.green(purpose)}\n`);
+        return { purpose: purpose as string, conversationHistory };
+      } else {
+        throw new Error(
+          'Failed to discover purpose: server did not call `done` when max turns are reached',
+        );
+      }
+    } else {
+      if (!question) {
+        throw new Error(`Failed to discover purpose: ${res.statusText}`);
+      }
 
-    throw new Error('Failed to discover purpose after exhausting maximum turns');
-  } finally {
-    // Ensure progress bar is always stopped, even if an exception occurs
-    if (pbar) {
-      pbar.stop();
```

Comment:
don't remove this it's causing the tests to stay open

Reply:
It's deleting the outputted purpose.